File: fuzz_lab44_rust_crypto_timing/crypto_target.rs
use std::env;
use std::fs::File;
use std::io::Read;
use std::process;
use std::time::Instant;

const SECRET_KEY: &[u8; 16] = b"SECRET_HMAC_KEY!";

fn non_constant_time_compare(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    for i in 0..a.len() {
        if a[i] != b[i] {
            return false;
        }
    }
    true
}

fn process_crypto(data: &[u8]) {
    if data.len() < 8 {
        return;
    }

    let magic = u32::from_le_bytes([data[0], data[1], data[2], data[3]]);
    if magic != 0x43525950 {
        return;
    }

    let mode = data[4];
    if mode == 0xCC {
        if data.len() >= 21 {
            let candidate = &data[5..21];
            let start = Instant::now();
            let is_match = non_constant_time_compare(candidate, SECRET_KEY);
            let _elapsed = start.elapsed();
            if is_match {
                eprintln!("[!] CRYPTO SIDE CHANNEL SINK HIT");
                process::abort();
            }
        }
    } else {
        println!("[*] Safe crypto operation executed");
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        process::exit(1);
    }

    let mut file = match File::open(&args[1]) {
        Ok(f) => f,
        Err(_) => process::exit(1),
    };

    let mut buffer = Vec::new();
    if file.read_to_end(&mut buffer).is_err() || buffer.len() > 4096 {
        process::exit(1);
    }

    process_crypto(&buffer);
}

File: fuzz_lab55_golang_crypto_subtle/crypto_target.go
package main

import (
	"encoding/binary"
	"fmt"
	"os"
	"time"
)

var secretMasterKey = []byte("GO_MASTER_KEY_32B_SECRET_TOKEN!")

func nonConstantTimeEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func processCrypto(data []byte) {
	if len(data) < 8 {
		return
	}
	magic := binary.LittleEndian.Uint32(data[0:4])
	if magic != 0x474F5342 {
		return
	}
	mode := data[4]

	if mode == 0xCC && len(data) >= 5+len(secretMasterKey) {
		candidate := data[5 : 5+len(secretMasterKey)]
		start := time.Now()
		matched := nonConstantTimeEqual(candidate, secretMasterKey)
		_ = time.Since(start)

		if matched {
			fmt.Fprintf(os.Stderr, "[!] CRYPTO TIMING LEAK SINK HIT\n")
			panic("CRYPTO_SUBTLE_TIMING_LEAK_SINK")
		}
	} else {
		fmt.Println("[*] Safe crypto hash verification executed")
	}
}

func main() {
	if len(os.Args) < 2 {
		os.Exit(1)
	}
	data, err := os.ReadFile(os.Args[1])
	if err != nil || len(data) > 4096 {
		os.Exit(1)
	}
	processCrypto(data)
}
